python/ray/serve/_private/version.py

```diff
@@ -18,14 +29,16 @@ def __init__(self, code_version: Optional[str], user_config: Optional[Any] = Non
             self.unversioned = False
             self.code_version = code_version
 
-        self.user_config = user_config
-        # TODO(simon): make this xlang compatible
-        if isinstance(user_config, bytes):
-            serialized_user_config = user_config
-        else:
-            serialized_user_config = str.encode(json.dumps(user_config, sort_keys=True))
-        self.user_config_hash = crc32(serialized_user_config)
-        self._hash = crc32(serialized_user_config + self.code_version.encode("utf-8"))
+        self.deployment_config: DeploymentConfig = deployment_config
+        self.ray_actor_options: Dict = ray_actor_options
+        self.compute_hashes()
```

Comment:
I assume we don't allow anyone to update the attribute, since it will change the hash value. I suggest to change the class DeploymentVersion into pydantic and freeze it after initialization. Or we raise error when anyone update the attribute.

Reply:
Would it become too heavyweight if we change it to pydantic? I'm in favor of making deployment version immutable, though.

Reply:
both sounds good to me, "freezing the class" makes the code less error-proned.

Reply:
Yeah, agreed. I'll probably look into this in a potential follow up PR.